quota-3.14/edquota.8
.TH EDQUOTA 8
.SH NAME
edquota \- edit user quotas
.SH SYNOPSIS
.B edquota
[
.B \-p
.I protoname
] [
.BR \-u \ |
.B \-g
] [
.B \-r
] [
.B \-F
.I format-name
] [
.B \-f
.I filesystem
]
.IR username .\|.\|.
.LP
.B edquota
[
.BR \-u \ |
.B \-g
] [
.B \-F
.I format-name
] [
.B \-f
.I filesystem
]
.B \-t
.LP
.B edquota
[
.BR \-u \ |
.B \-g
] [
.B \-F
.I format-name
] [
.B \-f
.I filesystem
]
.B \-T
.IR username \ |
.IR groupname .\|.\|.
.SH DESCRIPTION
.IX  "edquota command"  ""  "\fLedquota\fP \(em edit user quotas"
.IX  edit "user quotas \(em \fLedquota\fP"
.IX  "user quotas"  "edquota command"  ""  "\fLedquota\fP \(em edit user quotas"
.IX  "disk quotas"  "edquota command"  ""  "\fLedquota\fP \(em edit user quotas"
.IX  "quotas"  "edquota command"  ""  "\fLedquota\fP \(em edit user quotas"
.IX  "filesystem"  "edquota command"  ""  "\fLedquota\fP \(em edit user quotas"
.B edquota
is a quota editor.  One or more users or groups may be specified on the command
line. If a number is given in the place of user/group name it is treated as
an UID/GID. For each user or group a temporary file is created with an
.SM ASCII
representation of the current disk quotas for that user or group and an editor
is then invoked on the file.  The quotas may then be modified, new
quotas added, etc.
Setting a quota to zero indicates that no quota should be imposed.
.PP
Users are permitted to exceed their soft limits for a grace period that
may be specified per filesystem.  Once the grace period has expired, the
soft limit is enforced as a hard limit.
.PP
The current usage information in the file is for informational purposes;
only the hard and soft limits can be changed.
.PP
Upon leaving the editor,
.B edquota
reads the temporary file and modifies the binary quota files to reflect
the changes made.
.LP
The editor invoked is
.BR editor (1)
unless either the
.SB EDITOR
or the
.SB VISUAL
environment variable specifies otherwise.
.LP
Only the super-user may edit quotas.
.SH OPTIONS
.TP
.B -r, --remote
Edit also non-local quota use rpc.rquotad on remote server to set quota.
This option is available only if quota tools were compiled with enabled
support for setting quotas over RPC.
The
.B -n
option is equivalent, and is maintained for backward compatibility.
.TP
.B -u, --user
Edit the user quota. This is the default.
.TP
.B -g, --group
Edit the group quota.
.TP
.B -p, --prototype=\f2protoname\f1
Duplicate the quotas of the prototypical user
specified for each user specified.  This is the normal
mechanism used to initialize quotas for groups of users.
.TP
.B -F, --format=\f2format-name\f1
Edit quota for specified format (ie. don't perform format autodetection).
Possible format names are:
.B vfsold
(version 1 quota),
.B vfsv0
(version 2 quota),
.B rpc
(quota over NFS),
.B xfs
(quota on XFS filesystem)
.TP
.B \-f, --filesystem \f2filesystem\f1
Perform specified operations only for given filesystem (default is to perform
operations for all filesystems with quota).
.TP
.B \-t, --edit-period
Edit the soft time limits for each filesystem.
In old quota format if the time limits are zero, the default time limits in
.B <linux/quota.h>
are used. In new quota format time limits must be specified (there is no default
value set in kernel). Time units of 'seconds', 'minutes', 'hours', and 'days'
are understood. Time limits are printed in the greatest possible time unit such that
the value is greater than or equal to one.
.TP
.B \-T, --edit-times
Edit time for the user/group when softlimit is enforced. Possible values
are 'unset' or number and unit. Units are the same as in
.B \-t
option.
.SH FILES
.PD 0
.TP 20
.BR aquota.user " or " aquota.group
quota file at the filesystem root (version 2 quota, non-XFS filesystems)
.TP
.BR quota.user " or " quota.group
quota file at the filesystem root (version 1 quota, non-XFS filesystems)
.TP
.B /etc/mtab
mounted filesystems table
.PD
.SH SEE ALSO
.BR quota (1),
.BR editor (1),
.BR quotactl (2),
.BR quotacheck (8),
.BR quotaon (8),
.BR repquota (8),
.BR setquota (8)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



quota-4.00~pre1/edquota.c
/*
 * Copyright (c) 1980, 1990 Regents of the University of California. All
 * rights reserved.
 * 
 * This code is derived from software contributed to Berkeley by Robert Elz at
 * The University of Melbourne.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met: 1. Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer. 2.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution. 3. All advertising
 * materials mentioning features or use of this software must display the
 * following acknowledgement: This product includes software developed by the
 * University of California, Berkeley and its contributors. 4. Neither the
 * name of the University nor the names of its contributors may be used to
 * endorse or promote products derived from this software without specific
 * prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

/*
 * Disk quota editor.
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <paths.h>
#include <stdlib.h>
#include <fcntl.h>
#include <getopt.h>

#include "pot.h"
#include "quotaops.h"
#include "quotasys.h"
#include "quotaio.h"
#include "common.h"

#define FL_EDIT_PERIOD 1
#define FL_EDIT_TIMES 2
#define FL_REMOTE 4
#define FL_NUMNAMES 8
#define FL_NO_MIXED_PATHS 16

char *progname;

int flags, quotatype;
int fmt = -1;
char *protoname;
char *dirname;

void usage(void)
{
#if defined(RPC_SETQUOTA)
	char *rpcflag = "[-rm] ";
#else
	char *rpcflag = "";
#endif
	errstr(_("Usage:\n\tedquota %1$s[-u] [-F formatname] [-p username] [-f filesystem] username ...\n\
\tedquota %1$s-g [-F formatname] [-p groupname] [-f filesystem] groupname ...\n\
\tedquota %1$s[-u|g] [-F formatname] [-f filesystem] -t\n\
\tedquota %1$s[-u|g] [-F formatname] [-f filesystem] -T username|groupname ...\n"), rpcflag);
	fputs(_("\n\
-u, --user                    edit user data\n\
-g, --group                   edit group data\n"), stderr);
#if defined(RPC_SETQUOTA)
	fputs(_("-r, --remote                  edit remote quota (via RPC)\n\
-m, --no-mixed-pathnames      trim leading slashes from NFSv4 mountpoints\n"), stderr);
#endif
	fputs(_("-F, --format=formatname       edit quotas of a specific format\n\
-p, --prototype=name          copy data from a prototype user/group\n\
    --always-resolve          always try to resolve name, even if it is\n\
                              composed only of digits\n\
-f, --filesystem=filesystem   edit data only on a specific filesystem\n\
-t, --edit-period             edit grace period\n\
-T, --edit-times              edit grace time of a user/group\n\
-h, --help                    display this help text and exit\n\
-V, --version                 display version information and exit\n\n"), stderr);
	fprintf(stderr, _("Bugs to: %s\n"), MY_EMAIL);
	exit(1);
}

int parse_options(int argc, char **argv)
{
	int ret;
	struct option long_opts[] = {
		{ "help", 0, NULL, 'h' },
		{ "version", 0, NULL, 'V' },
		{ "prototype", 1, NULL, 'p' },
		{ "user", 0, NULL, 'u' },
		{ "group", 0, NULL, 'g' },
		{ "format", 1, NULL, 'F' },
		{ "filesystem", 1, NULL, 'f' },
#if defined(RPC_SETQUOTA)
		{ "remote", 0, NULL, 'r' },
		{ "no-mixed-pathnames", 0, NULL, 'm' },
#endif
		{ "always-resolve", 0, NULL, 256 },
		{ "edit-period", 0, NULL, 't' },
		{ "edit-times", 0, NULL, 'T' },
		{ NULL, 0, NULL, 0 }
	};

	if (argc < 2)
		usage();

	quotatype = USRQUOTA;
#if defined(RPC_SETQUOTA)
	while ((ret = getopt_long(argc, argv, "ugrmntTVp:F:f:", long_opts, NULL)) != -1) {
#else
	while ((ret = getopt_long(argc, argv, "ugtTVp:F:f:", long_opts, NULL)) != -1) {
#endif
		switch (ret) {
		  case 'p':
			  protoname = optarg;
			  break;
		  case 'g':
			  quotatype = GRPQUOTA;
			  break;
#if defined(RPC_SETQUOTA)
		  case 'n':
		  case 'r':
			  flags |= FL_REMOTE;
			  break;
		  case 'm':
			  flags |= FL_NO_MIXED_PATHS;
			  break;
#endif
		  case 'u':
			  quotatype = USRQUOTA;
			  break;
		  case 't':
			  flags |= FL_EDIT_PERIOD;
			  break;
		  case 'T':
			  flags |= FL_EDIT_TIMES;
			  break;
		  case 'F':
			  if ((fmt = name2fmt(optarg)) == QF_ERROR)	/* Error? */
				  exit(1);
			  break;
		  case 'f':
			  dirname = optarg;
			  break;
		  case 256:
			  flags |= FL_NUMNAMES;
			  break;
		  case 'V':
			  version();
			  exit(0);
		  default:
			  usage();
		}
	}
	argc -= optind;

	if (((flags & FL_EDIT_PERIOD) && argc != 0) || ((flags & FL_EDIT_TIMES) && argc < 1))
		usage();
	if ((flags & (FL_EDIT_PERIOD | FL_EDIT_TIMES)) && protoname) {
		errstr(_("Prototype name does not make sense when editing grace period or times.\n"));
		usage();
	}
	return optind;
}

void copy_prototype(int argc, char **argv, struct quota_handle **handles)
{
	int ret, protoid, id;
	struct dquot *protoprivs, *curprivs, *pprivs, *cprivs;
	
	ret = 0;
	protoid = name2id(protoname, quotatype, !!(flags & FL_NUMNAMES), NULL);
	protoprivs = getprivs(protoid, handles, 0);
	while (argc-- > 0) {
		id = name2id(*argv, quotatype, !!(flags & FL_NUMNAMES), NULL);
		curprivs = getprivs(id, handles, 0);
		if (!curprivs)
			die(1, _("Cannot get quota information for user %s\n"), *argv);
		argv++;

		for (pprivs = protoprivs, cprivs = curprivs; pprivs && cprivs;
		     pprivs = pprivs->dq_next, cprivs = cprivs->dq_next) {
			if (!devcmp_handles(pprivs->dq_h, cprivs->dq_h)) {
				errstr(_("fsname mismatch\n"));
				continue;
			}
			cprivs->dq_dqb.dqb_bsoftlimit =
				pprivs->dq_dqb.dqb_bsoftlimit;
			cprivs->dq_dqb.dqb_bhardlimit =
				pprivs->dq_dqb.dqb_bhardlimit;
			cprivs->dq_dqb.dqb_isoftlimit =
				pprivs->dq_dqb.dqb_isoftlimit;
			cprivs->dq_dqb.dqb_ihardlimit =
				pprivs->dq_dqb.dqb_ihardlimit;
			update_grace_times(cprivs);
		}
		if (putprivs(curprivs, COMMIT_LIMITS) == -1)
			ret = -1;
		freeprivs(curprivs);
	}
	if (dispose_handle_list(handles) == -1)
		ret = -1;
	freeprivs(protoprivs);
	exit(ret ? 1 : 0);
}

int main(int argc, char **argv)
{
	struct dquot *curprivs;
	int tmpfd, ret, id;
	struct quota_handle **handles;
	char *tmpfil, *tmpdir = NULL;

	gettexton();
	progname = basename(argv[0]);
	ret = parse_options(argc, argv);
	argc -= ret;
	argv += ret;

	init_kernel_interface();
	handles = create_handle_list(dirname ? 1 : 0, dirname ? &dirname : NULL, quotatype, fmt,
			(flags & FL_NO_MIXED_PATHS) ? 0 : IOI_NFS_MIXED_PATHS,
			(flags & FL_REMOTE) ? 0 : MS_LOCALONLY);
	if (!handles[0]) {
		dispose_handle_list(handles);
		fputs(_("No filesystems with quota detected.\n"), stderr);
		return 0;
	}
	if (protoname)
		copy_prototype(argc, argv, handles);

	umask(077);
	if (getuid() == geteuid() && getgid() == getegid())
		tmpdir = getenv("TMPDIR");
	if (!tmpdir)
		tmpdir = _PATH_TMP;
	tmpfil = smalloc(strlen(tmpdir) + strlen("/EdP.aXXXXXX") + 1);
	strcpy(tmpfil, tmpdir);
	strcat(tmpfil, "/EdP.aXXXXXX");
	tmpfd = mkstemp(tmpfil);
	if (tmpfd < 0) {
		errstr(_("Cannot create temporary file: %s\n"), strerror(errno));
		ret = -1;
		goto out;
	}
	if (fchown(tmpfd, getuid(), getgid()) < 0) {
		errstr(_("Cannot change owner of temporary file: %s\n"), strerror(errno));
		ret = -1;
		goto out;
	}
	ret = 0;
	if (flags & FL_EDIT_PERIOD) {
		if (writetimes(handles, tmpfd) < 0) {
			errstr(_("Cannot write grace times to file.\n"));
			ret = -1;
		}
		if (editprivs(tmpfil) < 0) {
			errstr(_("Error while editing grace times.\n"));
			ret = -1;
		}
		if (readtimes(handles, tmpfd) < 0) {
			errstr(_("Failed to parse grace times file.\n"));
			ret = -1;
		}
	}
	else if (flags & FL_EDIT_TIMES) {
		for (; argc > 0; argc--, argv++) {
			id = name2id(*argv, quotatype, !!(flags & FL_NUMNAMES), NULL);
			curprivs = getprivs(id, handles, 0);
			if (!curprivs)
				die(1, _("Cannot get quota information for user %s.\n"), *argv);
			if (writeindividualtimes(curprivs, tmpfd, *argv, quotatype) < 0) {
				errstr(_("Cannot write individual grace times to file.\n"));
				ret = -1;
				continue;
			}
			if (editprivs(tmpfil) < 0) {
				errstr(_("Error while editing individual grace times.\n"));
				ret = -1;
				continue;
			}
			if (readindividualtimes(curprivs, tmpfd) < 0) {
				errstr(_("Cannot read individual grace times from file.\n"));
				ret = -1;
				continue;
			}
			if (putprivs(curprivs, COMMIT_TIMES) == -1)
				ret = -1;
			freeprivs(curprivs);
		}
	}
	else {
		for (; argc > 0; argc--, argv++) {
			id = name2id(*argv, quotatype, !!(flags & FL_NUMNAMES), NULL);
			curprivs = getprivs(id, handles, 0);
			if (!curprivs)
				die(1, _("Cannot get quota information for user %s.\n"), *argv);
			if (writeprivs(curprivs, tmpfd, *argv, quotatype) < 0) {
				errstr(_("Cannot write quotas to file.\n"));
				ret = -1;
				continue;
			}
			if (editprivs(tmpfil) < 0) {
				errstr(_("Error while editing quotas.\n"));
				ret = -1;
				continue;
			}
			close(tmpfd);
			if ((tmpfd = open(tmpfil, O_RDONLY)) < 0)
				die(1, _("Cannot reopen!"));
			if (readprivs(curprivs, tmpfd) < 0) {
				errstr(_("Cannot read quotas from file.\n"));
				ret = -1;
				continue;
			}
			if (putprivs(curprivs, COMMIT_LIMITS) == -1)
				ret = -1;
			freeprivs(curprivs);
		}
	}
out:
	if (dispose_handle_list(handles) == -1)
		ret = -1;

	close(tmpfd);
	unlink(tmpfil);
	free(tmpfil);
	return ret ? 1 : 0;
}
